### ISSUE TITLE
"Detecting hardware encoders..." does not work in AppImage on Pop!OS

### ISSUE THREAD

**Issue by probonopd** [mltframework/shotcut#760]
The Pop!OS 19.04 Live ISO comes with NVIDIA Driver version 418.56 preinstalled, running from the Live system. Still, Shotcut cannot seem to be able to make use of hardware acceleration for encoding:

```
[Info   ] <Application::Application> Starting Shotcut version 19.07.15 
[Info   ] <Application::Application> Linux version 
[Info   ] <Application::Application> number of logical cores = 2 
[Info   ] <Application::Application> locale = QLocale(C, Default, Default) 
[Info   ] <Application::Application> install dir = "/tmp/.mount_ShotcunuwhdY/usr/bin/bin" 
(...)
[Info   ] <MainWindow::showStatusMessage> "Detecting hardware encoders..." 
[Info   ] <EncodeDock::on_hwencodeCheckBox_clicked> checking for "h264_nvenc" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "/tmp/.mount_ShotcunuwhdY/usr/bin/bin/ffmpeg" ("-hide_banner", "-f", "lavfi", "-i", "color=s=640x360", "-t", "0.040", "-an", "-c:v", "h264_nvenc", "-f", "rawvideo", "pipe:") 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Input #0, lavfi, from 'color=s=640x360':" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "  Duration: N/A, start: 0.000000, bitrate: N/A" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "    Stream #0:0: Video: rawvideo (I420 / 0x30323449), yuv420p, 640x360 [SAR 1:1 DAR 16:9], 25 tbr, 25 tbn, 25 tbc" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Stream mapping:" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "  Stream #0:0 -> #0:0 (rawvideo (native) -> h264 (h264_nvenc))" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Press [q] to stop, [?] for help" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "[h264_nvenc @ 0x1d99d00] Cannot init CUDA" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Conversion failed!" 
[Info   ] <EncodeDock::on_hwencodeCheckBox_clicked> checking for "hevc_nvenc" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "/tmp/.mount_ShotcunuwhdY/usr/bin/bin/ffmpeg" ("-hide_banner", "-f", "lavfi", "-i", "color=s=640x360", "-t", "0.040", "-an", "-c:v", "hevc_nvenc", "-f", "rawvideo", "pipe:") 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Input #0, lavfi, from 'color=s=640x360':" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "  Duration: N/A, start: 0.000000, bitrate: N/A" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "    Stream #0:0: Video: rawvideo (I420 / 0x30323449), yuv420p, 640x360 [SAR 1:1 DAR 16:9], 25 tbr, 25 tbn, 25 tbc" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Stream mapping:" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "  Stream #0:0 -> #0:0 (rawvideo (native) -> hevc (hevc_nvenc))" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Press [q] to stop, [?] for help" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "[hevc_nvenc @ 0x19e8d00] Cannot init CUDA" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Conversion failed!" 
[Info   ] <EncodeDock::on_hwencodeCheckBox_clicked> checking for "h264_vaapi" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "/tmp/.mount_ShotcunuwhdY/usr/bin/bin/ffmpeg" ("-hide_banner", "-f", "lavfi", "-i", "color=s=640x360", "-t", "0.040", "-an", "-vaapi_device", ":0", "-vf", "format=nv12,hwupload", "-c:v", "h264_vaapi", "-f", "rawvideo", "pipe:") 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: VA-API version 0.39.0" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: va_getDriverName() returns 0" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: Trying to open /tmp/.mount_ShotcunuwhdY/usr/bin/lib/va/nvidia_drv_video.so" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: va_openDriver() returns -1" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "[AVHWDeviceContext @ 0xe12fc0] Failed to initialise VAAPI connection: -1 (unknown libva error)." 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Device creation failed: -5." 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Failed to set value ':0' for option 'vaapi_device': Input/output error" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Error parsing global options: Input/output error" 
[Info   ] <EncodeDock::on_hwencodeCheckBox_clicked> checking for "hevc_vaapi" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "/tmp/.mount_ShotcunuwhdY/usr/bin/bin/ffmpeg" ("-hide_banner", "-f", "lavfi", "-i", "color=s=640x360", "-t", "0.040", "-an", "-vaapi_device", ":0", "-vf", "format=nv12,hwupload", "-c:v", "hevc_vaapi", "-f", "rawvideo", "pipe:") 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: VA-API version 0.39.0" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: va_getDriverName() returns 0" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: Trying to open /tmp/.mount_ShotcunuwhdY/usr/bin/lib/va/nvidia_drv_video.so" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: va_openDriver() returns -1" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "[AVHWDeviceContext @ 0x20ccfc0] Failed to initialise VAAPI connection: -1 (unknown libva error)." 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Device creation failed: -5." 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Failed to set value ':0' for option 'vaapi_device': Input/output error" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Error parsing global options: Input/output error" 
[Info   ] <MainWindow::showStatusMessage> "Nothing found" 

pop-os@pop-os:~$ cat /etc/os-release
NAME="Pop!_OS"
VERSION="19.04"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Pop!_OS 19.04"
VERSION_ID="19.04"
HOME_URL="https://system76.com/pop"
SUPPORT_URL="http://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=disco
UBUNTU_CODENAME=disco
```

**Comment by ddennedy** [comment]
I confirmed Pop! with NVIDIA does have libcuda installed. I do not think this can be fixed without making a build based on a newer Ubuntu OS (we use 16.04), which will decrease compatibility. There must be something version-dependent at compile time with respect to NVENC or CUDA. It is difficult to support hardware encoding with prebuilt binaries for Linux due to versioning and fragmentation. My target for Linux is primarily the last two Ubuntu LTS and derivatives/cousins thereof.

**Comment by probonopd** [comment]
I understand the graphics driver situation on Linux is a big mess; however it would be nice to have at least one combination of Live ISO and Shotcut AppImage that would work together. Thanks for having a look!

**Comment by ddennedy** [comment]
More test results with this Live ISO: hardware encode through VAAPI works on my machine with AMD Radeon.
Here is the ffmpeg debug level log output from the probe/test:
```
[h264_nvenc @ 0xf2c780] Loaded lib: libcuda.so.1
[h264_nvenc @ 0xf2c780] Loaded sym: cuInit
[h264_nvenc @ 0xf2c780] Loaded sym: cuDeviceGetCount
[h264_nvenc @ 0xf2c780] Loaded sym: cuDeviceGet
[h264_nvenc @ 0xf2c780] Loaded sym: cuDeviceGetAttribute
[h264_nvenc @ 0xf2c780] Loaded sym: cuDeviceGetName
[h264_nvenc @ 0xf2c780] Loaded sym: cuDeviceComputeCapability
[h264_nvenc @ 0xf2c780] Loaded sym: cuCtxCreate_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuCtxSetLimit
[h264_nvenc @ 0xf2c780] Loaded sym: cuCtxPushCurrent_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuCtxPopCurrent_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuCtxDestroy_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuMemAlloc_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuMemAllocPitch_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuMemsetD8Async
[h264_nvenc @ 0xf2c780] Loaded sym: cuMemFree_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuMemcpy2D_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuMemcpy2DAsync_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuGetErrorName
[h264_nvenc @ 0xf2c780] Loaded sym: cuGetErrorString
[h264_nvenc @ 0xf2c780] Loaded sym: cuStreamCreate
[h264_nvenc @ 0xf2c780] Loaded sym: cuStreamQuery
[h264_nvenc @ 0xf2c780] Loaded sym: cuStreamSynchronize
[h264_nvenc @ 0xf2c780] Loaded sym: cuStreamDestroy_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuStreamAddCallback
[h264_nvenc @ 0xf2c780] Loaded sym: cuEventCreate
[h264_nvenc @ 0xf2c780] Loaded sym: cuEventDestroy_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuEventSynchronize
[h264_nvenc @ 0xf2c780] Loaded sym: cuEventQuery
[h264_nvenc @ 0xf2c780] Loaded sym: cuEventRecord
[h264_nvenc @ 0xf2c780] Loaded sym: cuLaunchKernel
[h264_nvenc @ 0xf2c780] Loaded sym: cuModuleLoadData
[h264_nvenc @ 0xf2c780] Loaded sym: cuModuleUnload
[h264_nvenc @ 0xf2c780] Loaded sym: cuModuleGetFunction
[h264_nvenc @ 0xf2c780] Loaded sym: cuTexObjectCreate
[h264_nvenc @ 0xf2c780] Loaded sym: cuTexObjectDestroy
[h264_nvenc @ 0xf2c780] Loaded sym: cuGLGetDevices_v2
[h264_nvenc @ 0xf2c780] Loaded sym: cuGraphicsGLRegisterImage
[h264_nvenc @ 0xf2c780] Loaded sym: cuGraphicsUnregisterResource
[h264_nvenc @ 0xf2c780] Loaded sym: cuGraphicsMapResources
[h264_nvenc @ 0xf2c780] Loaded sym: cuGraphicsUnmapResources
[h264_nvenc @ 0xf2c780] Loaded sym: cuGraphicsSubResourceGetMappedArray
[h264_nvenc @ 0xf2c780] Loaded sym: cuDeviceGetUuid
[h264_nvenc @ 0xf2c780] Loaded sym: cuImportExternalMemory
[h264_nvenc @ 0xf2c780] Loaded sym: cuDestroyExternalMemory
[h264_nvenc @ 0xf2c780] Loaded sym: cuExternalMemoryGetMappedBuffer
[h264_nvenc @ 0xf2c780] Loaded sym: cuExternalMemoryGetMappedMipmappedArray
[h264_nvenc @ 0xf2c780] Loaded sym: cuMipmappedArrayGetLevel
[h264_nvenc @ 0xf2c780] Loaded sym: cuMipmappedArrayDestroy
[h264_nvenc @ 0xf2c780] Loaded sym: cuImportExternalSemaphore
[h264_nvenc @ 0xf2c780] Loaded sym: cuDestroyExternalSemaphore
[h264_nvenc @ 0xf2c780] Loaded sym: cuSignalExternalSemaphoresAsync
[h264_nvenc @ 0xf2c780] Loaded sym: cuWaitExternalSemaphoresAsync
[h264_nvenc @ 0xf2c780] Loaded lib: libnvidia-encode.so.1
[h264_nvenc @ 0xf2c780] Loaded sym: NvEncodeAPICreateInstance
[h264_nvenc @ 0xf2c780] Loaded sym: NvEncodeAPIGetMaxSupportedVersion
[h264_nvenc @ 0xf2c780] Loaded Nvenc version 9.0
[h264_nvenc @ 0xf2c780] Nvenc initialized successfully
[h264_nvenc @ 0xf2c780] Cannot init CUDA
[h264_nvenc @ 0xf2c780] Nvenc unloaded
```
A comment I read said I may need to load the nvidia_uvm kernel module, and indeed it was not loaded. I loaded via modprobe and confirm it loaded, but that did not help.

[Here](https://devtalk.nvidia.com/default/topic/1052054/linux/ffmpeg-cannot-init-cuda-for-transcoding/) is a good trouble-shooting thread. 
```
pop-os@pop-os:~$ systemctl status nvidia-persistenced.service
● nvidia-persistenced.service - NVIDIA Persistence Daemon
   Loaded: loaded (/lib/systemd/system/nvidia-persistenced.service; static; vendor preset: enabled)
   Active: inactive (dead)
$ systemctl start nvidia-persistenced.service
$ systemctl status nvidia-persistenced.service
● nvidia-persistenced.service - NVIDIA Persistence Daemon
   Loaded: loaded (/lib/systemd/system/nvidia-persistenced.service; static; vendor preset: enabled)
   Active: active (running) since Sun 2019-07-21 04:44:19 UTC; 22s ago
  Process: 5838 ExecStart=/usr/bin/nvidia-persistenced --user nvidia-persistenced --no-persistence-mode --verbose 
 Main PID: 5847 (nvidia-persiste)
    Tasks: 1 (limit: 4915)
   Memory: 916.0K
   CGroup: /system.slice/nvidia-persistenced.service
           └─5847 /usr/bin/nvidia-persistenced --user nvidia-persistenced --no-persistence-mode --verbose

Jul 21 04:44:19 pop-os systemd[1]: Starting NVIDIA Persistence Daemon...
Jul 21 04:44:19 pop-os nvidia-persistenced[5847]: Verbose syslog connection opened
Jul 21 04:44:19 pop-os nvidia-persistenced[5847]: Now running with user ID 116 and group ID 123
Jul 21 04:44:19 pop-os nvidia-persistenced[5847]: Started (5847)
Jul 21 04:44:19 pop-os nvidia-persistenced[5847]: device 0000:04:00.0 - registered
Jul 21 04:44:19 pop-os nvidia-persistenced[5847]: Local RPC services initialized
Jul 21 04:44:19 pop-os systemd[1]: Started NVIDIA Persistence Daemon.
```
Now it works!
I am not certain if it is necessary to modprobe nvidia_uvm. Hmm... after I stop the nvidia-persistenced service, it still works. But after `sudo rmmod nvidia_uvm` it stops working again. After this a `modprobe nvidia_uvm` lets it work again without nvidia-persistenced running.

In summary, one must run `sudo modprobe nvidia_uvm`. It might also be required to run `systemctl start nvidia-persistenced.service` to initialize something with CUDA in the driver.

**Comment by ddennedy** [comment]
No, it is simpler than that. Somewhere along the way, I ran a test ffmpeg command with sudo, and that triggered it into working. A reboot and test confirms that all you need to do is
1. Start Shotcut with sudo.
2. In Export, next to **Use hardware encoder** click **Configure...** It succeeds (assuming supporting hardware, of course).
3. Now hardware encoding works as the normal user as well.

**Comment by probonopd** [comment]
Thank you very mucvh @ddennedy. This seems to get nvenc going, but not libva/vaapi.
The whole situation is a bit [confusing](https://devtalk.nvidia.com/default/topic/1033137/linux/whats-the-situation-with-vdpau-vaapi-nvdec-/), but I guess I should be happy with nvenc on NVIDIA hardware.

```
Info   ] <EncodeDock::on_hwencodeCheckBox_clicked> checking for "hevc_vaapi" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "/tmp/.mount_Shotcu9E4ZFy/usr/bin/bin/ffmpeg" ("-hide_banner", "-f", "lavfi", "-i", "color=s=640x360", "-t", "0.040", "-an", "-vaapi_device", ":0", "-vf", "format=nv12,hwupload", "-c:v", "hevc_vaapi", "-f", "rawvideo", "pipe:") 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: VA-API version 0.39.0" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: va_getDriverName() returns 0" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: Trying to open /tmp/.mount_Shotcu9E4ZFy/usr/bin/lib/va/nvidia_drv_video.so" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "libva info: va_openDriver() returns -1" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "[AVHWDeviceContext @ 0x778fc0] Failed to initialise VAAPI connection: -1 (unknown libva error)." 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Device creation failed: -5." 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Failed to set value ':0' for option 'vaapi_device': Input/output error" 
[Debug  ] <EncodeDock::on_hwencodeCheckBox_clicked> "Error parsing global options: Input/output error" 
```
